### PR TITLE
Refactor to application

### PR DESCRIPTION
**Pull request by xifle**
Noisytest is based on some test python scripts primarily intended to test the suitability of the ML approach for 
system testing. As part of this PR, the scripts are refactored to a usable and complete application.

### CODE REVIEW

File: noisytest/experiment_reader.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+class ExperimentReader:
+    """Reads in experiment (training or validation) data"""
+
+    def __init__(self, import_preprocessor):
+        """Construct from ImportPreprocessor instance. The import_preprocessor is used
+        to generate a tensor from raw training / validation data
+        """
+        self._preprocessor = import_preprocessor
+        self._do_pad_data = True
```
Comment (krlx) [inline noisytest/experiment_reader.py]: Make do_pad_data settable via __init__ ? (with `True` als default)

File: noisytest/dataset_reader.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import glob
+import os
+
+import noisytest.experiment_reader
+
+
+class DataSetReader(noisytest.ExperimentReader):
+    """Reader for a whole dataset consisting of multiple experiments in one folder"""
+
+    def read_data_set(self, path):
```
Comment (krlx) [inline noisytest/dataset_reader.py:10]: Sometimes you have type hints (e.g., `def do_pad_data(self, value: bool)`), sometimes not like here

File: noisytest/experiment_reader.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    def read_experiment(self, data_path: str, experiment_name: str):
+        """Reads in an experiment consisting of noise data and labels"""
+
+        full_filename = os.path.join(data_path, experiment_name)
+
+        with noisytest.NoiseReader(full_filename) as noise_reader:
+            noise_data = noise_reader.data()
+            assert self._preprocessor.is_matching_sample_time(noise_reader.sample_time), \
+                "%r: Sample time does not match preprocessor specs" % full_filename
```
Comment (krlx) [inline noisytest/experiment_reader.py]: Assert should be used to ensure internal correctnes/as a debugging aid, raise an exception instead for user/configuration mistakes

File: noisytest/experiment_reader.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        if end_time == -1:
+            end_time = time_data[-1] + 0.1
+
+        block_duration = end_time - start_time
+        if block_duration < self._preprocessor.frame_duration and not self.do_pad_data:
+            # Don't pad - use a broader chunk than specified instead
+            start_time -= (self._preprocessor.frame_duration - block_duration) / 2.0
+            end_time = start_time + self._preprocessor.frame_duration + 1.0 / self._preprocessor.input_data_sample_rate
+
+        assert start_time < end_time, "%r: start time for a block must be less than end time" % filename
```
Comment (krlx) [inline noisytest/experiment_reader.py]: Raise Exception instead (if you are very motivated, use your own custom exception classes ;) )

File: noisytest/meta_data_reader.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    def __init__(self, filename: str):
+        """Construct a meta data reader from given filename
+
+        The file extension is automatically appended if none is supplied
+        """
+        if not filename.endswith(self.file_extension()):
+            filename = filename + self.file_extension()
+
+        self._reader = toml.load(filename)
```
Comment (krlx) [inline noisytest/meta_data_reader.py:26]: toml library opens/closes file for you (https://github.com/uiri/toml/blob/master/toml/decoder.py) -> no need to use as context manager?

File: noisytest/noise_reader.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        """Construct a noise data reader from given filename
+
+        The file extension is automatically appended if none is supplied
+        """
+        self._preprocessor = preprocessor
+
+        if not filename.endswith(self.file_extension()):
+            filename = filename + self.file_extension()
+
+        self._file_handle = open(filename, newline='')
```
Comment (krlx) [inline noisytest/noise_reader.py]: `with open('filename') as file:`
(Advantage is that file is properly closed, even if an exception is raised at some point.)

File: noisytest/optimizer.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        if self._set_hyper_parameter_on(self._pipeline.model, name, value):
+            return
+
+        processor = self._pipeline.feature_preprocessor
+        while processor is not None:
+            if self._set_hyper_parameter_on(processor, name, value):
+                return
+            processor = processor.parent
+
+        assert False, "Internal error"
```
Comment (krlx) [inline noisytest/optimizer.py]: Hyperparamter not found

File: noisytest/optimizer.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        """Recursively search the space of given hyper_parameters with a line search"""
+
+        if not hyper_parameters:
+            self._evaluate()
+            return
+
+        sub_space_parameters = copy.copy(hyper_parameters)
+        param = sub_space_parameters.popitem()
+
+        for value in param[1]:
```
Comment (krlx) [inline noisytest/optimizer.py]: What is param[1]??? :D

File: noisytest/optimizer.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            self._set_hyper_parameter(param[0], value)
+            self._search_sub_space(sub_space_parameters)
+
+    def grid_search(self):
+        """Performs a grid search on all hyper parameters of the pipeline"""
+        hyper_params = self._find_hyper_parameters()
+
+        self._search_sub_space(hyper_params)
+
+        print("Best fitness is", self._best_fitness)
```
Comment (krlx) [inline noisytest/optimizer.py]: Is it on purpose that sometime you use print, sometimes the logger?

File: noisytest/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        return self._model.validate(self._feature_preprocessor.prepare_input_target_data(validation_data))
+
+    def optimize(self, training_data, validation_data):
+        opt = noisytest.Optimizer(training_data, validation_data, self)
+        opt.grid_search()
+
+    def create_import_preprocessor(self, **kwargs):
+        return noisytest.TimeDataFramer(**kwargs)
+
+    def create_feature_preprocessor(self, **kwargs):
```
Comment (krlx) [inline noisytest/pipeline.py:75]: Maybe it would be nice to have to have a docstring explaining what is done here - you have one below (`"""Load noise data from given file"""`) where it is relatively clear what is done there ;)

File: noisytest/pipeline.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        return noisytest.Flatten(noisytest.DiscreteCosineTransform(
+            noisytest.Mag2Log(noisytest.SpectrogramCompressor(noisytest.Spectrogram(**kwargs)))))
+
+    def create_model(self, **kwargs):
+        return noisytest.Model(**kwargs)
+
+    def load_data(self, filename: str):
```
Comment (krlx) [inline noisytest/pipeline.py]: `load_training_data` below -> `load_prediction_data` or `load_target_data` here? To make it clearer

File: noisytest/pipeline.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+        """Loads data from specified directory using the import preprocessor
+
+        Returns a tuple (training_data, validation_data)
+        """
+        reader = noisytest.DataSetReader(self._import_preprocessor)
+        reader.do_pad_data = True
+
+        training_data = reader.read_data_set(training_data_directory)
+
+        reader.do_pad_data = False
```
Comment (krlx) [inline noisytest/pipeline.py:101]: Reference to paper/short explanation, why padding=True for train, =False for val

File: noisytest/preprocessor.py [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+    @property
+    def target_data_to_keywords(self):
+        """Maps target data to training data keywords"""
+        return dict((v, k) for k, v in self.keywords_to_target_data.items())
+
+    def append_parent(self, new_parent):
+        """Add a parent to the preprocessor chain"""
+        obj = self
+        while True:
+            if obj.parent is None:
```
Comment (krlx) [inline noisytest/preprocessor.py]: `if not obj.parent`

File: noisytest/preprocessor.py [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+    def prepare_data(self, data):
+        result = super().prepare_data(data)
+        result = self._frame_data(result)
+        return result
+
+    def prepare_input_target_data(self, data):
+        result = super().prepare_input_target_data(data)
+        result = self._pad_if_necessary(result)
+
+        assert self.keywords_to_target_data, "LabelData must be available"
```
Comment (krlx) [inline noisytest/preprocessor.py]: Raise Exception ;)

File: noisytest/preprocessor.py [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+        return abs(1.0 / sample_time - self.input_data_sample_rate) < self.input_data_sample_rate / 10.0
+
+    @property
+    def keywords_to_target_data(self):
+        """Maps training data keywords to target data"""
+        return self._keywords_to_target_data
+
+    @property
+    def hyper_parameters(self):
+        return {}
```
Comment (krlx) [inline noisytest/preprocessor.py:217]: is this necessary? Can't be set?

File: noisytest/preprocessor.py [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+
+    def process(self, data):
+        return tf.matmul(data, self._averaging_tensor(tf.shape(data)[-1], self.compression_factor))
+
+    @staticmethod
+    def _averaging_tensor(original_size, compression_rate):
+        """A tensor to compress the last dimension of tensor data by compression_rate"""
+        if original_size % compression_rate > 0:
+            print('originalSize:', original_size, 'compressionRate:', compression_rate)
+            raise ValueError('original size must be a multiple of the compression rate!')
```
Comment (krlx) [inline noisytest/preprocessor.py]: 👍 an Exception

**Review by krlx — COMMENTED** [review]
add README -> LGTM ;)
